### PR TITLE
Macros

### PR DESCRIPTION
**Pull request by rpiaggio**
A more detailed documentation is pending, but in summary this PR implements 2 annotation macros:

* `GraphQLSchema`, which generates an object with types for all `enum`s  and `input`s from a schema, allowing for manual overrides.
* `GraphQL` which generates types to hold the results of an operation + variables + convenience call method.

Operations and clients are now typed with a schema (which can be any type, but they must match).

It's not 100% clear that we will end up using these annotation macros (particularly since dotty yet has to support them). However, programming using the macros results in an Apollo-like experience, dramatically reducing boilerplate to handle responses. I expect that whatever approach we end up using (we may end up generating code or maybe even using a compiler plugin?), the client code will be pretty much the same as what is achieved with these macros. Therefore, I think it's worthwhile to merge and start using it in client projects.

Please note that a huge chunk of the size of the PR is because of the inclusion of several schemas used for testing.

### CODE REVIEW

File: macros/src/main/scala/clue/macros/GraphQL.scala [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+    macroResolve(annottees: _*)
+
+  /**
+   * Extract the `document` contents from the `GraphQLQuery` the marcro was applied to.
+   */
+  @tailrec
+  private[this] def documentDef(tree: List[Tree]): Option[String] =
+    tree match {
+      case Nil                                          => none
+      case q"$mods val document: $tpe = $document" :: _ =>
```
Comment (cquiroz) [inline macros/src/main/scala/clue/macros/GraphQL.scala:40]: is the content of a quasiquote compiled?
Can you put illegal code inside it?

File: macros/src/main/scala/clue/macros/GraphQL.scala [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        Select(TypeNamesToTermNames.transform(schemaType), TermName(scope)),
+        List(ImportSelector(termNames.WILDCARD, -1, null, -1))
+      ),
+      Apply(Ident(TermName(s"ignoreUnusedImport$scope")), List.empty)
+    )
+
+  private[this] def addImports(schemaType: Tree): List[Tree] => List[Tree] =
+    parentBody => {
+      val importDef = schemaImport(schemaType) _
+      // val importDef = Import(
```
Comment (cquiroz) [inline macros/src/main/scala/clue/macros/GraphQL.scala:155]: Commented code

**Review by cquiroz — APPROVED** [review]
I'm totally unqualified to review macros but it surely looks impressive.
The final result is worth it

File: macros/src/main/scala/clue/macros/GraphQL.scala (hunk 1 above)
Comment (rpiaggio) [reply macros/src/main/scala/clue/macros/GraphQL.scala:40]: Quasiquotes translate into ASTs. They do have to express a valid AST or the macro won't compile. That said, nothing prevents you from building an illegal AST.

File: macros/src/main/scala/clue/macros/GraphQL.scala (hunk 2 above)
Comment (rpiaggio) [reply macros/src/main/scala/clue/macros/GraphQL.scala:155]: Thank you!

**Review by swalker2m — APPROVED** [review]
I fear I don't have anything constructive to offer here.  I concur with Carlos that it looks amazing, but I don't really feel qualified to review it.  I'll second his approval at least.

**Pull request merged**